README
see COPYING for the license (LGPL)

Dependencies
============
- Reasonably compliant C++ compiler (gcc 3.x, vc 7.1, etc)
- Boost library: See www.boost.org for details and downloads
- CppUnit (optional for test suite): See cppunit.sf.net
- (optional) SConsAddons - Addon library for the build system.  https://realityforge.vrsource.org/SconsAddons

Building
========
CppDom uses the scons build system.  You can download scons from www.scons.org.

We also use the SconsAddons library (see above).  It is included in the source
tree or you can add its source directory in your PYTHONPATH to build.

ex: https://realityforge.vrsource.org/svn/scons-addons/trunk/scons-addons

PYTHONPATH=/home/username/scons-addons/src

NOTE: In the future we hope to use third party branches or svn:externals to 
remove the need to download this separately.

Once you have both of these, the follow steps should be all you need.

Please see 'scons --help' for a list of options for specifying where the
dependencies are located.

type 'scons' to build cppdom shared lib.
type 'scons install' to install on the system
type 'scons install prefix=$HOME/software' to install in your account

Windows
=======

CppDOM can be compiled with Visual C++ 6.0 and newer. However, SCons tries to
be smart and pick the version of Visual C++ that it thinks you should use
(always--or almost always--the newest version). To override this, use the
MSVS_VERSION command line option. For example, to force compilation with
Visual C++ 7.0 (from Visual Studio .NET 2002), run the following:

   scons MSVS_VERSION=7.0

Let's say we want to use Visual C++ 8.0 (from Visual Studio 8) instead. In
that case, run the following:

   scons MSVS_VERSION=8.0

If there is only one version of Visual Studio installed, SCons will find it,
and using the MSVS_VERSION argument is unnecessary.

64-bit Windows
==============

To build a 64-bit version of CppDOM on Windows, Visual Studio 2005 (also
known as Visual Studio 8) must be used, and it must be run on a 64-bit
Windows installation. The trick, however, is that the 32-bit version of
Python must be used. Otherwise, SCons will not find the Visual Studio
installation because it looks in the wrong place in the Registry and in
"C:\Program Files\Microsoft Visual Studio 8" instead of in
"C:\Program Files (x86)\Microsoft Visual Studio 8".

So, to build both the 32-bit and the 64-bit versions of CppDOM, run the
following:

   scons MSVS_VERSION=8.0

To build only the 64-bit version, run the following:

   scons MSVS_VERSION=8.0 var_arch=x64

Mac OS X
========

On Mac OS X, the build may auto-detect that it can build multiple
architectures (not necessarily universal binaries), but this case is not
currently handled well. SCons will complain that there are multiple ways to
build the same target and exit with an error. To work around this, specify
the desired architecture explicitly (pass var_arch=ppc, var_arch=ia32, or
var_arch=ppc64 on the scons command line) or build a universal binary (see
below).

Universal Binaries
==================

To build the CppDOM libraries as universal binaries (Mac OS X only), use the
following command:

   scons darwin_universal=yes var_arch=universal var_type=optimized \
         darwin_sdk=/Developer/SDKs/MacOSX10.4u.sdk

An alternate platform SDK can be used by using the appropriate value with
the darwin_sdk command line arcument.


ChangeLog
DATE        AUTHOR      CHANGE
---------- -----------  -------------------------------------------------------
2012-07-04 patrickh     Add CMake build.
                        Developed by Ryan Pavlik.

[Version 1.2.0 released - 5.1.2011]============================================

2011-04-23 patrickh     SCons 2.0 is now the minimum required version.
                        Submitted by Doug McCorkle.
2010-04-14 patrickh     Preserve newlines in CDATA.
                        Submitted by: Carsten Neumann
                        VERSION: 1.1.2
2010-03-06 patrickh     Fixed conflicts with the operator<< overload for
                        std::ostream and implicit constructio of
                        cppdom::Attribute objects.
                        Submitted by: Carsten Neumann
                        VERSION: 1.1.1

[Version 1.0.0 released - 3.2.2009]============================================

2007-08-01 patrickh     Updated for SConsAddons changes. On Windows, the
                        build formerly called "hybrid" is now the "debug"
                        build. The DLL is installed in lib\debug. The build
                        formerly called "debug" is now the "debugrt" build (to
                        indicate that it is linked against the Visual C++ debug
                        runtime). The library is still named with "_d".
                        VERSION: 0.7.10
2007-07-25 patrickh     Fixed empty comment processing bug.
2007-07-06 patrickh     Fixed build issues on Mac OS X.
                        VERSION: 0.7.9
2007-06-26 patrickh     Made many build fixes and improvements on Windows.
                        VERSION: 0.7.8
2006-11-07 patrickh     Do not put headers into a versioned directory on
                        Windows.
                        VERSION: 0.7.7
2006-07-13 patrickh     Implemented automatic linking on Windows.
                        VERSION: 0.7.6
2006-07-13 patrickh     Moved the declaration of cppdom::getVersion() to
                        cppdom.h.
2006-07-11 allenb       Expose OptionRepository and cppdom methods in dll.
2006-07-05 allenb       More refactoring of build system.
                        It is not possible to specify all variants to build
                        and build them simultaneously.
2006-06-30 allenb       Refatored the build system to support building on Win32
                        using SCons.  This requires an in-development version
                        of scons-addons to get the EnvironmentBuilder.
2006-06-24 patrickh     Support building universal binaries on Mac OS X.
2006-06-13 allenb       Fixed bug in how NodePtr and DocumentPtr types were
                        defined.  Now explicitly declare Node and Document.
                        VERSION: 0.7.5
2006-06-06 dshipton     Version out includes and library
                        VERSION: 0.7.4
2006-06-05 allenb       Refactored the build system to remove AutoDist.
2006-06-02 dshipton     Install .pc file for pkg-config use
                        VERSION: 0.7.3
2006-02-07 allenb       Refactored the Attributes class to just be an std::map
                        of strings to attributes.
                        Added an attrib() method to Node that returns this map.
                        VERSION: 0.7.2
2006-05-06 allenb       Added ability to set options in OptionRepository.
                        Added saveOptionFile to OptionRepository.
                        Refactored OptionRepository to hold onto Doc node.
                        cppdom::Error now derived from std::exception
                        VERSION: 0.7.1
                        
[Version 0.6.0 released - 6.1.2005]============================================
2005-02-06 allenb       Added hasOption() method to OptionRepository
                        VERSION: 0.5.6
2005-02-04 allenb       Added ability to get children by path.
                        Added extension to assist with using xml files
                        as configuration files.  See: ext/OptionRepository.h
                        VERSION: 0.5.5
2005-02-03 allenb       Added tests for Node interface.
                        Added Node::hasChild() method.
                        VERSION: 0.5.4
2004-04-14 allenb       Added ability to make distribution as part of build.
                        VERSION: 0.5.3
2004-04-14 allenb       Migrated build system to use AutoDist.
                        VERSION: 0.5.2
2004-04-13 allenb       Refactoring build system to use scons-addons
                        VERSION: 0.5.1
2004-04-12 allenb       Moving into active development version again
                        NEW VERSION: 0.5.0

[Version 0.4.0 released - 4.6.2004]============================================
2004-02-29 allenb       Moved the node type enum into Node as Node::Type. 
                        This breaks all old code that used the type interface.
                        To help avoid future breakage in the future, added
                        an interface to Node to query for node type. (isLeaf(),
                        isNode(), isCData(), isDocument()).
2004-02-25 allenb       Modified build to take into account separate boost
                        include directory.
2004-02-21 allenb       Started adding tests for boost spirit parsing.
                        Added options for boost spirit
                        NEW VERSION: 0.3.1
2004-02-20 allenb       Added more extensive tests to test suite.  Added metric
                        tests for loading and parsing documents.
2004-02-14 allenb       Restructure the test suite.  Added StaticOnly option
                        to build.  Added Metric Registry to test suite.

[Version 0.3.0 released - 1.28.2004]===========================================
2004-01-28 patrickh     Restored the ability to compile shared libraries.
2003-10-07 allenb       Added support for "stringification" of attrib text and

include/omniORB4/IOP.h
// -*- Mode: C++; -*-
//                            Package   : omniORB
// IOP.h                      Created on: 8/2/96
//                            Author    : Sai Lai Lo (sll)
//
//    Copyright (C) 1996-1999 AT&T Laboratories Cambridge
//
//    This file is part of the omniORB library
//
//    The omniORB library is free software; you can redistribute it and/or
//    modify it under the terms of the GNU Library General Public
//    License as published by the Free Software Foundation; either
//    version 2 of the License, or (at your option) any later version.
//
//    This library is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
//    Library General Public License for more details.
//
//    You should have received a copy of the GNU Library General Public
//    License along with this library; if not, write to the Free
//    Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  
//    02111-1307, USA
//
//
// Description:
//       C++ mapping of the OMG IOP module
//       Reference: CORBA V2.0 10.6.2
//	
//	

/*
  $Log: IOP.h,v $
  Revision 1.4.2.4  2007/11/28 12:24:26  dgrisby
  Implement a tiny subset of CSIv2 to permit multiple SSL endpoints in IORs.

  Revision 1.4.2.3  2006/07/18 16:21:24  dgrisby
  New experimental connection management extension; ORB core support
  for it.

  Revision 1.4.2.2  2005/01/06 23:08:09  dgrisby
  Big merge from omni4_0_develop.

  Revision 1.4.2.1  2003/03/23 21:04:20  dgrisby
  Start of omniORB 4.1.x development branch.

  Revision 1.2.2.8  2001/08/06 15:49:16  sll
  Added IOP component TAG_OMNIORB_UNIX_TRANS for omniORB specific local
  transport using the unix domain socket.

  Revision 1.2.2.7  2001/07/31 16:10:39  sll
  Added GIOP BiDir support.

  Revision 1.2.2.6  2001/04/18 17:52:19  sll
  Added FT specification constants.

  Revision 1.2.2.5  2000/11/15 17:00:55  sll
  Added marshalling operators for ServiceContext.

  Revision 1.2.2.4  2000/11/03 18:58:46  sll
  Unbounded sequence of octet got a new type name.

  Revision 1.2.2.3  2000/10/04 16:52:04  sll
  New helper function dumpComponent().

  Revision 1.2.2.2  2000/09/27 17:06:09  sll
  Updated to use the new cdrStream abstraction.
  Updated to include the latest componentID, service context ID etc.

  Revision 1.2.2.1  2000/07/17 10:35:34  sll
  Merged from omni3_develop the diff between omni3_0_0_pre3 and omni3_0_0.

  Revision 1.3  2000/07/13 15:26:05  dpg1
  Merge from omni3_develop for 3.0 release.

  Revision 1.1.2.1  1999/09/24 09:51:41  djr
  Moved from omniORB2 + some new files.

  Revision 1.7  1999/06/18 21:13:56  sll
  Updated copyright notice.

  Revision 1.6  1999/06/18 20:36:05  sll
  Replaced _LC_attr with _core_attr.

  Revision 1.5  1999/01/07 18:22:25  djr
  Replaced _OMNIORB_NTDLL_IMPORT with _LC_attr.

  Revision 1.4  1998/04/07 19:59:00  sll
   Replace _OMNIORB2_NTDLL_ specification on class IOP with
  _OMNIORB_NTDLL_IMPORT on static member constants.

  Revision 1.3  1997/12/09 20:36:28  sll
  Added TaggedProfileList_var class.

 * Revision 1.2  1997/05/06  16:07:17  sll
 * Public release.
 *
 */

#ifndef __OMNIORB_IOP_H__
#define __OMNIORB_IOP_H__

class IOP {
public:
  
  typedef _CORBA_ULong ProfileId;
  static _core_attr const   ProfileId    TAG_INTERNET_IOP;
  static _core_attr const   ProfileId    TAG_MULTIPLE_COMPONENTS;
  static _core_attr const   ProfileId    TAG_SCCP_IOP;

  struct TaggedProfile {
    ProfileId tag;
    _CORBA_Unbounded_Sequence_Octet profile_data;

    void operator>>= (cdrStream &s) const;
    void operator<<= (cdrStream &s);
  };

  typedef _CORBA_Unbounded_Sequence<TaggedProfile> TaggedProfileList;

  class TaggedProfileList_out;
  class TaggedProfileList_var {
  public:
    typedef TaggedProfileList _Tseq;
    typedef TaggedProfileList_var _T_var;

    inline TaggedProfileList_var() : pd_seq(0) {}
    inline TaggedProfileList_var(_Tseq* s) : pd_seq(s) {}
    inline TaggedProfileList_var(const _T_var& sv) {
      if( sv.pd_seq ) {
	pd_seq = new _Tseq;
	*pd_seq = *sv.pd_seq;
      } else
	pd_seq = 0;
    }
    inline ~TaggedProfileList_var() { if( pd_seq ) delete pd_seq; }

    inline _T_var& operator = (_Tseq* s) {
      if( pd_seq )  delete pd_seq;
      pd_seq = s;
      return *this;
    }
    inline _T_var& operator = (const _T_var& sv) {
      if( sv.pd_seq ) {
	if( !pd_seq )  pd_seq = new _Tseq;
	*pd_seq = *sv.pd_seq;
      } else if( pd_seq ) {
	delete pd_seq;
	pd_seq = 0;
      }
      return *this;
    }

    inline TaggedProfile& operator [] (_CORBA_ULong i) {   return (*pd_seq)[i]; }
    inline _Tseq* operator -> () { return pd_seq; }
#if defined(__GNUG__) && __GNUG__ == 2 && __GNUC_MINOR__ == 7
    inline operator _Tseq& () const { return *pd_seq; }
#else
    inline operator const _Tseq& () const { return *pd_seq; }
    inline operator _Tseq& () { return *pd_seq; }
#endif

    inline const _Tseq& in() const { return *pd_seq; }
    inline _Tseq& inout() { return *pd_seq; }
    inline _Tseq*& out() { if (pd_seq) { delete pd_seq; pd_seq = 0; } return pd_seq; }
    inline _Tseq* _retn() { _Tseq* tmp = pd_seq; pd_seq = 0; return tmp; }

    friend class TaggedProfileList_out;

  private:
    _Tseq* pd_seq;
  };


  class TaggedProfileList_out {
  public:
    typedef TaggedProfileList _Tseq;
    typedef TaggedProfileList_var _T_var;

    inline TaggedProfileList_out(_Tseq*& s) : _data(s) { _data = 0; }
    inline TaggedProfileList_out(_T_var& sv)
      : _data(sv.pd_seq) { sv = (_Tseq*) 0; }
    inline TaggedProfileList_out(const TaggedProfileList_out& s) : _data(s._data) { }
    inline TaggedProfileList_out& operator=(const TaggedProfileList_out& s) { _data = s._data; return *this; }
    inline TaggedProfileList_out& operator=(_Tseq* s) { _data = s; return *this; }
    inline operator _Tseq*&() { return _data; }
    inline _Tseq*& ptr() { return _data; }
    inline _Tseq* operator->() { return _data; }
    inline TaggedProfile& operator [] (_CORBA_ULong i) {   return (*_data)[i]; }
    _Tseq*& _data;

  private:
    TaggedProfileList_out();
    TaggedProfileList_out operator=( const _T_var&);
  };

  struct IOR {
    _CORBA_String_member  type_id;
    TaggedProfileList     profiles;

    // the following are omniORB private functions
    void operator>>= (cdrStream &s);
    void operator<<= (cdrStream &s);
    static char* unmarshaltype_id(cdrStream&);
  };

  typedef _CORBA_ULong ComponentId;
  static _core_attr const ComponentId TAG_ORB_TYPE;
  static _core_attr const ComponentId TAG_CODE_SETS;
  static _core_attr const ComponentId TAG_POLICIES;
  static _core_attr const ComponentId TAG_ALTERNATE_IIOP_ADDRESS;
  static _core_attr const ComponentId TAG_ASSOCIATION_OPTIONS;
  static _core_attr const ComponentId TAG_SEC_NAME;
  static _core_attr const ComponentId TAG_SPKM_1_SEC_MECH;
  static _core_attr const ComponentId TAG_SPKM_2_SEC_MECH;
  static _core_attr const ComponentId TAG_KERBEROSV5_SEC_MECH;
  static _core_attr const ComponentId TAG_CSI_ECMA_SECRET_SEC_MECH;
  static _core_attr const ComponentId TAG_CSI_ECMA_HYBRID_SEC_MECH;
  static _core_attr const ComponentId TAG_SSL_SEC_TRANS;
  static _core_attr const ComponentId TAG_CSI_ECMA_PUBLIC_SEC_MECH;
  static _core_attr const ComponentId TAG_GENERIC_SEC_MECH;
  static _core_attr const ComponentId TAG_FIREWALL_TRANS;
  static _core_attr const ComponentId TAG_SCCP_CONTACT_INFO;
  static _core_attr const ComponentId TAG_JAVA_CODEBASE;
  static _core_attr const ComponentId TAG_CSI_SEC_MECH_LIST;
  static _core_attr const ComponentId TAG_NULL_TAG;
  static _core_attr const ComponentId TAG_TLS_SEC_TRANS;
  static _core_attr const ComponentId TAG_COMPLETE_OBJECT_KEY;
  static _core_attr const ComponentId TAG_ENDPOINT_ID_POSITION;
  static _core_attr const ComponentId TAG_LOCATION_POLICY;
  static _core_attr const ComponentId TAG_DCE_STRING_BINDING;
  static _core_attr const ComponentId TAG_DCE_BINDING_NAME;
  static _core_attr const ComponentId TAG_DCE_NO_PIPES;
  static _core_attr const ComponentId TAG_DCE_SEC_MECH;
  static _core_attr const ComponentId TAG_INET_SEC_TRANS;
  static _core_attr const ComponentId TAG_GROUP; // FT SPEC
  static _core_attr const ComponentId TAG_PRIMARY; // FT SPEC
  static _core_attr const ComponentId TAG_HEARTBEAT_ENABLED; // FT SPEC

  // omniORB specific:
  static _core_attr const ComponentId TAG_OMNIORB_BIDIR;
  static _core_attr const ComponentId TAG_OMNIORB_UNIX_TRANS;
  static _core_attr const ComponentId TAG_OMNIORB_PERSISTENT_ID;
  static _core_attr const ComponentId TAG_OMNIORB_RESTRICTED_CONNECTION;


  static const char* ComponentIDtoName(ComponentId);
  // omniORB private function.
  // Return the name given the ComponentId. Return nil if the ComponentId
  // is not recongised.

  struct TaggedComponent {
    ComponentId	            tag;
    _CORBA_Unbounded_Sequence_Octet component_data;

    // the following are omniORB private functions
    void operator>>= (cdrStream &s) const;
    void operator<<= (cdrStream &s);
  };
  
  static char* dumpComponent(const TaggedComponent&);
  // omniORB private function. Produce the textual dump of the component
  // content

  typedef _CORBA_Unbounded_Sequence<TaggedComponent> MultipleComponentProfile;

  class MultipleComponentProfile_out;

  class MultipleComponentProfile_var {
  public:
    typedef MultipleComponentProfile _Tseq;
    typedef MultipleComponentProfile_var _T_var;

    inline MultipleComponentProfile_var() : pd_seq(0) {}
    inline MultipleComponentProfile_var(_Tseq* s) : pd_seq(s) {}
    inline MultipleComponentProfile_var(const _T_var& sv) {
      if( sv.pd_seq ) {
	pd_seq = new _Tseq;
	*pd_seq = *sv.pd_seq;
      } else
	pd_seq = 0;
    }
    inline ~MultipleComponentProfile_var() { if( pd_seq ) delete pd_seq; }

    inline _T_var& operator = (_Tseq* s) {
      if( pd_seq )  delete pd_seq;
      pd_seq = s;
      return *this;
    }
    inline _T_var& operator = (const _T_var& sv) {
      if( sv.pd_seq ) {
	if( !pd_seq )  pd_seq = new _Tseq;
	*pd_seq = *sv.pd_seq;
      } else if( pd_seq ) {
	delete pd_seq;
	pd_seq = 0;
      }
      return *this;
    }

    inline TaggedComponent& operator [] (_CORBA_ULong i) {   return (*pd_seq)[i]; }
    inline _Tseq* operator -> () { return pd_seq; }
#if defined(__GNUG__) && __GNUG__ == 2 && __GNUC_MINOR__ == 7
    inline operator _Tseq& () const { return *pd_seq; }
#else
    inline operator const _Tseq& () const { return *pd_seq; }
    inline operator _Tseq& () { return *pd_seq; }
#endif

    inline const _Tseq& in() const { return *pd_seq; }
    inline _Tseq& inout() { return *pd_seq; }
    inline _Tseq*& out() { if (pd_seq) { delete pd_seq; pd_seq = 0; } return pd_seq; }
    inline _Tseq* _retn() { _Tseq* tmp = pd_seq; pd_seq = 0; return tmp; }

    friend class MultipleComponentProfile_out;

  private:
    _Tseq* pd_seq;
  };

  class MultipleComponentProfile_out {
  public:
    typedef MultipleComponentProfile _Tseq;
    typedef MultipleComponentProfile_var _T_var;

    inline MultipleComponentProfile_out(_Tseq*& s) : _data(s) { _data = 0; }
    inline MultipleComponentProfile_out(_T_var& sv)
      : _data(sv.pd_seq) { sv = (_Tseq*) 0; }
    inline MultipleComponentProfile_out(const MultipleComponentProfile_out& s) : _data(s._data) { }
    inline MultipleComponentProfile_out& operator=(const MultipleComponentProfile_out& s) { _data = s._data; return *this; }
    inline MultipleComponentProfile_out& operator=(_Tseq* s) { _data = s; return *this; }
    inline operator _Tseq*&() { return _data; }
    inline _Tseq*& ptr() { return _data; }
    inline _Tseq* operator->() { return _data; }
    inline TaggedComponent& operator [] (_CORBA_ULong i) {   return (*_data)[i]; }
    _Tseq*& _data;

  private:
    MultipleComponentProfile_out();
    MultipleComponentProfile_out operator=( const _T_var&);
  };


  typedef _CORBA_Unbounded_Sequence<MultipleComponentProfile> MultipleComponentProfileList;

  typedef _CORBA_ULong ServiceID;

  struct ServiceContext {
    ServiceID              context_id;
    _CORBA_Unbounded_Sequence_Octet context_data;

    // the following are omniORB private functions
    void operator>>= (cdrStream &s) const;
    void operator<<= (cdrStream &s);
  };

  typedef _CORBA_Unbounded_Sequence<ServiceContext> ServiceContextList;

  class ServiceContextList_out;

  class ServiceContextList_var {
  public:
    typedef ServiceContextList _Tseq;
    typedef ServiceContextList_var _T_var;

    inline ServiceContextList_var() : pd_seq(0) {}
    inline ServiceContextList_var(_Tseq* s) : pd_seq(s) {}
    inline ServiceContextList_var(const _T_var& sv) {
      if( sv.pd_seq ) {
	pd_seq = new _Tseq;
	*pd_seq = *sv.pd_seq;
      } else
	pd_seq = 0;
    }
    inline ~ServiceContextList_var() { if( pd_seq ) delete pd_seq; }

    inline _T_var& operator = (_Tseq* s) {
      if( pd_seq )  delete pd_seq;
      pd_seq = s;
      return *this;
    }
    inline _T_var& operator = (const _T_var& sv) {
      if( sv.pd_seq ) {
	if( !pd_seq )  pd_seq = new _Tseq;
	*pd_seq = *sv.pd_seq;
      } else if( pd_seq ) {
	delete pd_seq;
	pd_seq = 0;
      }
      return *this;
    }

    inline ServiceContext& operator [] (_CORBA_ULong i) {   return (*pd_seq)[i]; }
    inline _Tseq* operator -> () { return pd_seq; }
#if defined(__GNUG__) && __GNUG__ == 2 && __GNUC_MINOR__ == 7
    inline operator _Tseq& () const { return *pd_seq; }
#else
    inline operator const _Tseq& () const { return *pd_seq; }
    inline operator _Tseq& () { return *pd_seq; }
#endif

    inline const _Tseq& in() const { return *pd_seq; }
    inline _Tseq& inout() { return *pd_seq; }
    inline _Tseq*& out() { if (pd_seq) { delete pd_seq; pd_seq = 0; } return pd_seq; }
    inline _Tseq* _retn() { _Tseq* tmp = pd_seq; pd_seq = 0; return tmp; }

    friend class ServiceContextList_out;

  private:
    _Tseq* pd_seq;
  };

  class ServiceContextList_out {
  public:
    typedef ServiceContextList _Tseq;
    typedef ServiceContextList_var _T_var;

    inline ServiceContextList_out(_Tseq*& s) : _data(s) { _data = 0; }
    inline ServiceContextList_out(_T_var& sv)
      : _data(sv.pd_seq) { sv = (_Tseq*) 0; }
    inline ServiceContextList_out(const ServiceContextList_out& s) : _data(s._data) { }
    inline ServiceContextList_out& operator=(const ServiceContextList_out& s) { _data = s._data; return *this; }
    inline ServiceContextList_out& operator=(_Tseq* s) { _data = s; return *this; }
    inline operator _Tseq*&() { return _data; }
    inline _Tseq*& ptr() { return _data; }
    inline _Tseq* operator->() { return _data; }
    inline ServiceContext& operator [] (_CORBA_ULong i) {   return (*_data)[i]; }
    _Tseq*& _data;

  private:
    ServiceContextList_out();
    ServiceContextList_out operator=( const _T_var&);
  };

  static _core_attr const ServiceID TransactionService;
  static _core_attr const ServiceID CodeSets; 
  static _core_attr const ServiceID ChainBypassCheck;
  static _core_attr const ServiceID ChainBypassInfo;
  static _core_attr const ServiceID LogicalThreadId;
  static _core_attr const ServiceID BI_DIR_IIOP;
  static _core_attr const ServiceID SendingContextRunTime;
  static _core_attr const ServiceID INVOCATION_POLICIES;
  static _core_attr const ServiceID FORWARDED_IDENTITY;
  static _core_attr const ServiceID UnknownExceptionInfo;
  static _core_attr const ServiceID RTCorbaPriority;
  static _core_attr const ServiceID RTCorbaPriorityRange;
  static _core_attr const ServiceID GROUP_VERSION; // FT SPEC
  static _core_attr const ServiceID REQUEST;       // FT SPEC

  static _core_attr const ServiceID OMNIORB_RESTRICTED_CONNECTION;

  static const char* ServiceIDtoName(ServiceID);
  // omniORB private function.
  // Return the name given the ComponentId. Return nil if the ComponentId
  // is not recongised.

};

#endif // __OMNIORB_IOP_H__
